### PR TITLE
.Net Bugfix: Azure SDK Retry Logic unaffected when Kernel Retry = 0

### LINKED ISSUE #2192: .Net: When using Azure Chat/Text completion, default retry settings prevent complete control over retry behaviour.
**Describe the bug**
When using the Azure extension method `.WithAzureChatCompletionService()` to add an instance of  `AzureChatCompletion` to the kernel, the methods/constructors currently do not expose the retry configuration for the underlying `Azure.AI.OpenAI.OpenAIClient`. This client defaults to a 3-retry policy, which is not the behavior we would like for our given application. Currently, the only way to control this, seems to be new-ing our own OpenAIClient, then use reflection to overwrite the underlying field the `OpenAIClient` object is stored in. This is obviously not a great production solution.

**To Reproduce**
1. Execute the SK setup below
2. Breach token limits so endpoint returns errors
3. Attempt to send a message while token limits are still breached
4. Observe client will sit and attempt 3 retries until the server gives a valid response.

**Expected behavior**
This should immediately fail when the underlying endpoint token rate has been breached, but the internal client has it's own retry mechanism.

SK Setup Bad, retries despite being disabled.
```csharp
HttpRetryConfig httpRetryConfig = new HttpRetryConfig() { MaxRetryCount = 0 };
DefaultHttpRetryHandlerFactory defaultHttpRetryHandlerFactory = new DefaultHttpRetryHandlerFactory(httpRetryConfig);
var kernel = new KernelBuilder()
     .WithAzureChatCompletionService(apiConfig.Deployment, apiConfig.Endpoint, apiConfig.ApiKey)
     .WithRetryHandlerFactory(defaultHttpRetryHandlerFactory)
     .Build();
```

**Platform**
 - OS: Windows
 - IDE: Visual Studio 2022
 - Language: C#
 - Source: NuGet

**Workaround**
SK Setup "Working" (Don't want to be using reflection if possible)
By taking the code from `.WithAzureChatCompletionService()` and adding some reflection, we can overwrite the OpenAI client being used, with a newly constructed one that has retries explicitly disabled.
```csharp
        AzureChatCompletion Factory((ILogger Logger, KernelConfig Config) parameters)
        {
            //Init the chat completion client just like default .WithAzureChatCompletionService().
            var handler = new AzureChatCompletion(
            apiConfig.Deployment,
            apiConfig.Endpoint,
            apiConfig.ApiKey,
            new HttpClient(),
            logger
            );

            var no_retry_options = new OpenAI.OpenAIClientOptions();
            no_retry_options.Retry.MaxRetries = 0;
            var no_retry_client = new OpenAI.OpenAIClient(new Uri(apiConfig.Endpoint), new AzureKeyCredential(apiConfig.ApiKey), no_retry_options);

            handler.GetType().GetProperty("Client", BindingFlags.NonPublic | BindingFlags.Instance)
                .GetAutomaticPropertyBackingField()
                .SetValue(handler, no_retry_client);

            return handler;
        }
```

### PR DESCRIPTION
**Pull request by RogerBarreto**
### Motivation and Context

According to the issue #2192 when using the Retry abstraction implementation with AzureText/Chat Completion the behavior is being ignored.

Azure Client adds the resiliency as part of the HttpPipeline while the SK resiliency is running in as a DelegateHttpHandler, this way even disabling the retry logic in the Delegate part, the internal Azure pipeline implementation kicks in with the default 3 retrials.

Closes #2192
Resolves #2192

### Description

Could identify that indeed the OpenAIClient implementation is actually using another approach for Retry logic through RetryPolicy or Retry property in the ClientOptions configuration.

As a mitigation for the bug I moved the extensions implementation to actually use another constructor which expect the OpenAIClient object and create its configuration as part of extension private methods impl with the same Retry policy configuration that the Kernel has.

Currently the changes don't break customers, but I'm still considering removing or changing the other constructors that don't make usage of `OpenAIClient` because if they are used directly (without thru the extension) the bug would still happen (Azure SDK pipeline defaulting to 3 retrials).

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Review by shawncal — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Reliability/DefaultHttpRetryHandlerFactory.cs [hunk 1]
```diff
@@ -14,8 +14,10 @@ public DefaultHttpRetryHandlerFactory(HttpRetryConfig? config = null)
 
     public DelegatingHandler Create(ILogger? logger)
     {
-        return new DefaultHttpRetryHandler(this._config, logger);
+        return new DefaultHttpRetryHandler(this.Config, logger);
     }
 
     private readonly HttpRetryConfig? _config;
+
+    public HttpRetryConfig? Config => this._config;
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/Reliability/DefaultHttpRetryHandlerFactory.cs]: Changes made, thanks.

**Review by shawncal — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/OpenAIClientBase.cs [hunk 2]
```diff
@@ -51,6 +51,26 @@ public abstract class OpenAIClientBase : ClientBase
...
+    /// Note: instances created this way might not have the default diagnostics settings,
+    /// it's up to the caller to configure the client.
+    /// </summary>
+    /// <param name="modelId">Azure OpenAI model ID or deployment name, see https://learn.microsoft.com/azure/cognitive-services/openai/how-to/create-resource</param>
+    /// <param name="openAIClient">Custom <see cref="OpenAIClient"/>.</param>
+    /// <param name="logger">Application logger</param>
+    private protected OpenAIClientBase(
+        string modelId,
+        OpenAIClient openAIClient,
+        ILogger? logger = null)
```
Comment (shawncal) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/AzureSdk/OpenAIClientBase.cs:65]: This constructor does nothing with logger

File: dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletion/OpenAIChatCompletion.cs [hunk 3]
```diff
@@ -35,6 +36,20 @@ public sealed class OpenAIChatCompletion : OpenAIClientBase, IChatCompletion, IT
...
+    /// Create an instance of the OpenAI chat completion connector
+    /// </summary>
+    /// <param name="modelId">Model name</param>
+    /// <param name="openAIClient">Custom <see cref="OpenAIClient"/> for HTTP requests.</param>
+    /// <param name="logger">Application logger</param>
+    public OpenAIChatCompletion(
+    string modelId,
+    OpenAIClient openAIClient,
+    ILogger? logger = null
+) : base(modelId, openAIClient, logger)
```
Comment (shawncal) [inline dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletion/OpenAIChatCompletion.cs:49]: Funny tabbing

**Comment by shawncal** [comment]
Note, a couple times the merge queue has rejected this due to integration test failures. It looks like a rate limiting issue, but odd that this is the only change impacted. Can you take a look at the failures please?
https://github.com/microsoft/semantic-kernel/actions/runs/5747067526/job/15577491707

**Comment by RogerBarreto** [comment]
> Note, a couple times the merge queue has rejected this due to integration test failures. It looks like a rate limiting issue, but odd that this is the only change impacted. Can you take a look at the failures please?

@shawncal Changed the Integration Tests to check for any exception message instead of being more specific, basically any failure would suffice to validate the test (no resiliency policy running in)

**Comment by shawncal** [comment]
Integration test failure: https://github.com/microsoft/semantic-kernel/actions/runs/5758216619/job/15610414061

SemanticKernel.IntegrationTests.Connectors.OpenAI.AzureOpenAICompletionTests.**AzureOpenAIChatNoHttpRetryPolicyCustomClientShouldThrowAsync**(prompt: "Where is the most famous fish market in Seattle, W"..., ???: "This model's maximum context length is") [FAIL]
  Skipped SemanticKernel.IntegrationTests.Connectors.Weaviate.WeaviateMemoryStoreTests.ItListsCollectionsAsync [1 ms]
  Skipped SemanticKernel.IntegrationTests.Connectors.OpenAI.OpenAITextEmbeddingTests.OpenAITestAsync [1 ms]
  Failed SemanticKernel.IntegrationTests.Connectors.OpenAI.AzureOpenAICompletionTests.**AzureOpenAIChatNoHttpRetryPolicyCustomClientShouldThrowAsync**(prompt: "Where is the most famous fish market in Seattle, W"..., ???: "This model's maximum context length is") [< 1 ms]
  Error Message:
   System.InvalidOperationException : The test method expected 1 parameter value, but 2 parameter values were provided.

File: dotnet/src/IntegrationTests/Connectors/OpenAI/AzureOpenAICompletionTests.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+        var exception = await Assert.ThrowsAsync<AIException>(() => func.InvokeAsync(string.Empty, settings: new CompleteRequestSettings() { MaxTokens = 1000000, Temperature = 0.5, TopP = 0.5 }));
+
+        // Assert
+        Assert.NotNull(exception);
+    }
+
+    [Theory]
+    [InlineData("Where is the most famous fish market in Seattle, Washington, USA?",
+        "This model's maximum context length is")]
```
Comment (shawncal) [inline dotnet/src/IntegrationTests/Connectors/OpenAI/AzureOpenAICompletionTests.cs:67]: ``
```suggestion
    [InlineData("Where is the most famous fish market in Seattle, Washington, USA?")]
```

**Review by shawncal — APPROVED** [review]

**Pull request merged**